// @vercel/cli-auth/credentials-store.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasCredentials = hasCredentials;
exports.resolveAuthTokenStorage = resolveAuthTokenStorage;
exports.getGlobalPathConfig = getGlobalPathConfig;
exports.getConfigFilePath = getConfigFilePath;
exports.getAuthConfigFilePath = getAuthConfigFilePath;
exports.readGlobalConfig = readGlobalConfig;
exports.getAuthTokenStorage = getAuthTokenStorage;
exports.setLoadKeyringModuleForTesting = setLoadKeyringModuleForTesting;
exports.CredentialsStore = CredentialsStore;
exports.readCredentials = readCredentials;
exports.writeCredentials = writeCredentials;
exports.clearCredentials = clearCredentials;
exports.clearAllCredentials = clearAllCredentials;
exports.clearAllCredentialsStrict = clearAllCredentialsStrict;
const node_path_1 = __importDefault(require("node:path"));
const node_crypto_1 = require("node:crypto");
const fs = __importStar(require("node:fs"));
const node_os_1 = require("node:os");
// NOTE: We are pinned on v5 because newer versions behave differently
// when there are dots in the application name, eg. `com.vercel.cli` becomes `com.vercel`
const xdg_app_paths_1 = __importDefault(require("xdg-app-paths"));
// Avoid exposing zod v4 types to TS 4.9 workspace consumers during type-check.
const { z } = require('zod/mini');
const AUTH_TOKEN_STORAGE = {
    AUTO: 'auto',
    FILE: 'file',
    KEYRING: 'keyring',
};
const AUTH_TOKEN_STORAGE_VALUES = Object.values(AUTH_TOKEN_STORAGE);
const DEFAULT_AUTH_TOKEN_STORAGE = AUTH_TOKEN_STORAGE.FILE;
const KEYRING_SERVICE = 'Vercel Auth';
const KEYRING_UNAVAILABLE_ERROR = 'OS keyring support is unavailable. Set `authTokenStorage` to `auto` or `file` to store credentials in auth.json instead.';
const defaultLoadKeyringModule = () => require('@napi-rs/keyring');
let loadKeyringModule = defaultLoadKeyringModule;
const CredentialsSchema = z.object({
    /** An `access_token` obtained using the OAuth Device Authorization flow. */
    token: z.optional(z.string()),
    /** The ID of the currently authenticated user, cached from `/v2/user`. */
    userId: z.optional(z.string()),
    /** A `refresh_token` obtained using the OAuth Device Authorization flow. */
    refreshToken: z.optional(z.string()),
    /**
     * The absolute time (seconds) when the {@link Credentials.token} expires.
     * Used to optimistically check if the token is still valid.
     */
    expiresAt: z.optional(z.number()),
    /** Whether to skip writing the credentials to disk during {@link CredentialsStore.update} */
    skipWrite: z.optional(z.boolean()),
    '// Note': z.optional(z.string()),
    '// Docs': z.optional(z.string()),
});
function hasCredentials(credentials) {
    return Boolean(credentials.token ||
        credentials.refreshToken ||
        typeof credentials.expiresAt === 'number');
}
function parseCredentials(value) {
    return CredentialsSchema.parse(value);
}
function isAuthTokenStorage(value) {
    return (typeof value === 'string' &&
        AUTH_TOKEN_STORAGE_VALUES.some(storage => storage === value));
}
function resolveAuthTokenStorage(value) {
    if (typeof value === 'undefined') {
        return DEFAULT_AUTH_TOKEN_STORAGE;
    }
    if (isAuthTokenStorage(value)) {
        return value;
    }
    throw new Error(`Invalid value for \`authTokenStorage\`: ${JSON.stringify(value)}. Expected one of: ${AUTH_TOKEN_STORAGE_VALUES.map(storage => JSON.stringify(storage)).join(', ')}.`);
}
/** Returns whether a directory exists */
function isDirectory(path) {
    try {
        return fs.lstatSync(path).isDirectory();
    }
    catch (_) {
        // We don't care which kind of error occured, it isn't a directory anyway.
        return false;
    }
}
/**
 * Returns in which directory the config should be present
 * @internal Should only be used in {@link CredentialsStore} or tests.
 */
function getGlobalPathConfig(appName) {
    const vercelDirectories = (0, xdg_app_paths_1.default)(appName).dataDirs();
    const possibleConfigPaths = [
        ...vercelDirectories, // latest vercel directory
        node_path_1.default.join((0, node_os_1.homedir)(), '.now'), // legacy config in user's home directory
        ...(0, xdg_app_paths_1.default)('now').dataDirs(), // legacy XDG directory
    ];
    // The customPath flag is the preferred location,
    // followed by the vercel directory,
    // followed by the now directory.
    // If none of those exist, use the vercel directory.
    return (possibleConfigPaths.find(configPath => isDirectory(configPath)) ||
        vercelDirectories[0]);
}
function getConfigFilePath(dir) {
    return node_path_1.default.join(dir, 'config.json');
}
function getAuthConfigFilePath(dir) {
    return node_path_1.default.join(dir, 'auth.json');
}
function readGlobalConfig(dir) {
    const configPath = getConfigFilePath(dir);
    let config;
    try {
        config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
    }
    catch (error) {
        if (isENOENTError(error)) {
            return {};
        }
        throw error;
    }
    return {
        authTokenStorage: typeof config.authTokenStorage === 'undefined'
            ? undefined
            : resolveAuthTokenStorage(config.authTokenStorage),
    };
}
function getAuthTokenStorage(dir, config) {
    if (config?.authTokenStorage) {
        return resolveAuthTokenStorage(config.authTokenStorage);
    }
    return resolveAuthTokenStorage(readGlobalConfig(dir).authTokenStorage);
}
function getKeyringAccount(configPath) {
    const digest = (0, node_crypto_1.createHash)('sha256').update(configPath).digest('hex');
    return `cli:${digest.slice(0, 16)}`;
}
function setLoadKeyringModuleForTesting(nextLoadKeyringModule) {
    loadKeyringModule = nextLoadKeyringModule ?? defaultLoadKeyringModule;
}
function createKeyringEntry(configPath) {
    try {
        const { Entry } = loadKeyringModule();
        return new Entry(KEYRING_SERVICE, getKeyringAccount(configPath));
    }
    catch (error) {
        const wrappedError = new Error(KEYRING_UNAVAILABLE_ERROR);
        wrappedError.cause = error;
        throw wrappedError;
    }
}
function isKeyringUnavailableError(error) {
    return error instanceof Error && error.message === KEYRING_UNAVAILABLE_ERROR;
}
function readCredentialsFromFile(configPath) {
    return parseCredentials(JSON.parse(fs.readFileSync(configPath, 'utf8')));
}
function isENOENTError(error) {
    return (typeof error === 'object' &&
        error !== null &&
        'code' in error &&
        error.code === 'ENOENT');
}
function createMissingCredentialsError(configPath) {
    const error = new Error(`ENOENT: no such file or directory, open '${configPath}'`);
    Object.assign(error, { code: 'ENOENT' });
    return error;
}
function writeCredentialsToFile(configPath, config) {
    const directory = node_path_1.default.dirname(configPath);
    const temporaryPath = node_path_1.default.join(directory, `${node_path_1.default.basename(configPath)}.${process.pid}.${(0, node_crypto_1.randomUUID)()}.tmp`);
    fs.mkdirSync(node_path_1.default.dirname(configPath), { recursive: true });
    try {
        fs.writeFileSync(temporaryPath, JSON.stringify(config, null, 2) + '\n', {
            mode: 0o600,
        });
        fs.renameSync(temporaryPath, configPath);
    }
    catch (error) {
        fs.rmSync(temporaryPath, { force: true });
        throw error;
    }
}
function readCredentialsFromKeyring(configPath) {
    const entry = createKeyringEntry(configPath);
    const value = entry.getPassword();
    if (value === null) {
        return null;
    }
    return parseCredentials(JSON.parse(value));
}
function writeCredentialsToKeyring(configPath, config) {
    const entry = createKeyringEntry(configPath);
    entry.setPassword(JSON.stringify(config));
}
function deleteCredentialsFromKeyring(configPath) {
    const entry = createKeyringEntry(configPath);
    return entry.deleteCredential();
}
function deleteCredentialsFile(configPath) {
    fs.rmSync(configPath, { force: true });
}
function tryDeleteCredentialsFromKeyring(configPath) {
    try {
        deleteCredentialsFromKeyring(configPath);
    }
    catch { }
}
function deleteCredentialsFromKeyringIfAvailable(configPath) {
    try {
        deleteCredentialsFromKeyring(configPath);
    }
    catch (error) {
        if (!isKeyringUnavailableError(error)) {
            throw error;
        }
    }
}
function migrateCredentialsToFile(configPath, credentials) {
    writeCredentialsToFile(configPath, credentials);
    tryDeleteCredentialsFromKeyring(configPath);
    return credentials;
}
function tryReadKeyringCredentialsIntoFile(configPath) {
    try {
        const credentials = readCredentialsFromKeyring(configPath);
        if (credentials === null) {
            return null;
        }
        return migrateCredentialsToFile(configPath, credentials);
    }
    catch {
        return null;
    }
}
function readFileStorageCredentials(configPath) {
    try {
        const credentials = readCredentialsFromFile(configPath);
        if (hasCredentials(credentials)) {
            return credentials;
        }
        // File mode treats keyring contents as stale-but-authoritative backup only
        // when auth.json has no usable credentials.
        return tryReadKeyringCredentialsIntoFile(configPath) ?? credentials;
    }
    catch (error) {
        if (!isENOENTError(error)) {
            throw error;
        }
        const credentials = tryReadKeyringCredentialsIntoFile(configPath);
        if (credentials !== null) {
            return credentials;
        }
        throw error;
    }
}
function migrateCredentialsToKeyring(configPath, credentials) {
    writeCredentialsToKeyring(configPath, credentials);
    deleteCredentialsFile(configPath);
    return credentials;
}
function readKeyringStorageCredentials(configPath) {
    const credentials = readCredentialsFromKeyring(configPath);
    if (credentials !== null) {
        return credentials;
    }
    try {
        const fileCredentials = readCredentialsFromFile(configPath);
        // Keyring mode eagerly migrates any remaining plaintext credentials out of
        // auth.json after the first successful read.
        return migrateCredentialsToKeyring(configPath, fileCredentials);
    }
    catch (error) {
        if (isENOENTError(error)) {
            throw createMissingCredentialsError(configPath);
        }
        throw error;
    }
}
function readAutoStorageCredentials(configPath) {
    try {
        const credentials = readCredentialsFromKeyring(configPath);
        if (credentials !== null) {
            return credentials;
        }
    }
    catch { }
    // Auto mode intentionally avoids read-time migration so that machines
    // without usable keyring support can keep working off auth.json.
    return readCredentialsFromFile(configPath);
}
function writeFileStorageCredentials(configPath, credentials) {
    writeCredentialsToFile(configPath, credentials);
    tryDeleteCredentialsFromKeyring(configPath);
    return 'file';
}
function writeKeyringStorageCredentials(configPath, credentials) {
    writeCredentialsToKeyring(configPath, credentials);
    deleteCredentialsFile(configPath);
    return 'keyring';
}
function writeAutoStorageCredentials(configPath, credentials) {
    try {
        return writeKeyringStorageCredentials(configPath, credentials);
    }
    catch {
        tryDeleteCredentialsFromKeyring(configPath);
        return writeFileStorageCredentials(configPath, credentials);
    }
}
function deleteFileStorageCredentials(configPath) {
    deleteCredentialsFile(configPath);
    deleteCredentialsFromKeyringIfAvailable(configPath);
}
function deleteKeyringStorageCredentials(configPath) {
    deleteCredentialsFromKeyring(configPath);
    deleteCredentialsFile(configPath);
}
function deleteAutoStorageCredentials(configPath) {
    try {
        deleteCredentialsFromKeyring(configPath);
    }
    catch { }
    deleteCredentialsFile(configPath);
}
function CredentialsStore(dir, config = {}) {
    const configPath = getAuthConfigFilePath(dir);
    const tokenStorage = getAuthTokenStorage(dir, config);
    return {
        configPath,
        get() {
            switch (tokenStorage) {
                case AUTH_TOKEN_STORAGE.FILE:
                    return readFileStorageCredentials(configPath);
                case AUTH_TOKEN_STORAGE.KEYRING:
                    return readKeyringStorageCredentials(configPath);
                case AUTH_TOKEN_STORAGE.AUTO:
                    return readAutoStorageCredentials(configPath);
            }
        },
        /** Update the credentials store. If `skipWrite` is set, the update will be skipped. */
        update(config) {
            if (config.skipWrite)
                return;
            const parsedConfig = parseCredentials(config);
            switch (tokenStorage) {
                case AUTH_TOKEN_STORAGE.FILE:
                    return writeFileStorageCredentials(configPath, parsedConfig);
                case AUTH_TOKEN_STORAGE.KEYRING:
                    return writeKeyringStorageCredentials(configPath, parsedConfig);
                case AUTH_TOKEN_STORAGE.AUTO:
                    return writeAutoStorageCredentials(configPath, parsedConfig);
            }
        },
        delete() {
            switch (tokenStorage) {
                case AUTH_TOKEN_STORAGE.FILE:
                    deleteFileStorageCredentials(configPath);
                    break;
                case AUTH_TOKEN_STORAGE.KEYRING:
                    deleteKeyringStorageCredentials(configPath);
                    break;
                case AUTH_TOKEN_STORAGE.AUTO:
                    deleteAutoStorageCredentials(configPath);
                    break;
            }
        },
    };
}
function readCredentials(dir, config = {}) {
    return CredentialsStore(dir, config).get();
}
function writeCredentials(dir, credentials, config = {}) {
    return CredentialsStore(dir, config).update(credentials);
}
function clearCredentials(dir, config = {}) {
    CredentialsStore(dir, config).delete();
}
function clearAllCredentials(dir) {
    for (const authTokenStorage of [
        AUTH_TOKEN_STORAGE.FILE,
        AUTH_TOKEN_STORAGE.KEYRING,
    ]) {
        try {
            clearCredentials(dir, { authTokenStorage });
        }
        catch { }
    }
}
function clearAllCredentialsStrict(dir) {
    const configPath = getAuthConfigFilePath(dir);
    let fileError;
    let keyringError;
    try {
        deleteCredentialsFile(configPath);
    }
    catch (error) {
        fileError = error;
    }
    try {
        deleteCredentialsFromKeyring(configPath);
    }
    catch (error) {
        if (!isKeyringUnavailableError(error)) {
            keyringError = error;
        }
    }
    if (keyringError) {
        throw keyringError;
    }
    if (fileError) {
        throw fileError;
    }
}
